Speed up sum over a dimension

colesbury opened the pull request:
Perf numbers:
https://gist.github.com/colesbury/9e28dd7b0f27b0b019f68adbd4bd4b88

I've changed the dispatch stub so that it doesn't require every kernel
to be compiled for every instruction set. Kernel implementations are
stored in the stub's table with the REGISTER_DISPATCH macro.

I've also moved vec256 to it's own folder and split up the
specializations before they get too unwieldy.

cpuhrsch commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -0,0 +1,51 @@
+  inline void load(const T *ptr) {
+    std::memcpy(values, ptr, 32);
+  };
+  static Vec256 s_load(const T* ptr) {
```
What does the s_ stand for?

colesbury replied:
`static`. I'm not tied to the name. You just can't have a static overload with the same name.

cpuhrsch commented on aten/src/ATen/cpu/vec256/vec256_int.h:
```diff
@@ -0,0 +1,121 @@
+#pragma once
```
You could also merge float and double into _floating just as with _int

Or we could also split on avx and avx2

colesbury replied:
It's split based on data type: e.g. `_m256i` vs. `_m256` vs. `_m256d`.

As a matter of practicality, it makes it easy to compare float vs. double and I expect these specializations to have the most functions added to it.

cpuhrsch commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -15,27 +15,23 @@
-
 // ALL REDUCE #################################################################
 
-Tensor _reduce_cpu(reduce_type *f, const Tensor &self) {
```
This will be nice to have once you add max, min, mean, etc.

cpuhrsch commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -73,22 +69,16 @@ static Tensor &_dimreduce_setup(Tensor &result, const Tensor &self,
   return result;
 }
 
-Tensor &_reduce_out_cpu(reduce_type *f, Tensor &result, const Tensor &self,
```
This will be nice to have once you add more ops

cpuhrsch commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -1,118 +1,154 @@
+}
+
+template<typename F>
+static void parallel_for(int64_t end, int64_t step, bool parallelize, F func) {
```
Why not put this into parallel.h?

colesbury replied:
init_tbb_num_threads is called by Reduction::apply which is the only entry point.

It's not in parallel.h because I'm not sure this pattern is generally useful as-is. If we start using it in another file then we can move it to a common header.

cpuhrsch commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -1,118 +1,154 @@
+
+    scalar_t sum;
+    if (size > internal::TBB_GRAIN_SIZE) {
+      sum = tbb::parallel_reduce(
```
This need to call the tbb init function as otherwise the scheduler isn't setup correctly, which is why all calls to tbb were aggregated in parallel.h/.cpp

colesbury replied:
init_tbb_num_threads is called by Reduction::apply which is the only entry point

cpuhrsch replied:
I'd add it simply as a convention, but I can't force you. The computational cost is negligible  and the cost of someone calling a tbb algorithm before the init justifies the extra line of code.

cpuhrsch commented:
Generally I'm not convinced the dispatch code is that much better, but I also don't care too much. I think we might to revisit it again in the future anyway. Having the ability to pass a table or function for non-standard cases is definitely good.

All calls to tbb will need to be moved into Parallel.h/.cpp or need to call the init_tbb_num_threads function. I'll work on making sure that tbb is not available to any file outside of Parallel.h as otherwise people will do this and we won't have exact control over the number of threads we use.

cpuhrsch commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -1,118 +1,154 @@
+  }
+}
+
+static tbb::affinity_partitioner ap;
```
If you're worried about this being per function, maybe we can make it thread-local and local to parallel.h/.cpp

colesbury replied:
I don't see an advantage of making this thread-local or per-function or putting it in parallel.h.

The reductions operations have the same access pattern. It makes sense to share an affinity partitioner between them.

colesbury committed:
Speed up sum over a dimension

Perf numbers:
https://gist.github.com/colesbury/9e28dd7b0f27b0b019f68adbd4bd4b88

I've changed the dispatch stub so that it doesn't require every kernel
to be compiled for every instruction set. Kernel implementations are
stored in the stub's table with the REGISTER_DISPATCH macro.

I've also moved vec256 to it's own folder and split up the
specializations before they get too unwieldy.

colesbury committed:
Change UnaryOpsKernel to use new DisaptchStub

 - Prefer signed integers. Mixing signed and unsigned integers is a
   pain and ATen mostly uses signed integers (int64_t).
 - Use inline lambda instead of struct for UnaryOps
 - Rename partial load overload "load_partial"

colesbury commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -1,76 +1,142 @@
+  });
+}
+
+static void sqrt_kernel(Tensor& result, const Tensor& self) {
```
These functions follow similar patterns, but they're not all exactly the same. For example, "abs" dispatches on integer as well as floating point types. Some of the unimplemented functions will likely change the body of the lambda.
